Doc/c-api/long.rst

```diff
@@ -442,6 +442,16 @@ distinguished from a number.  Use :c:func:`PyErr_Occurred` to disambiguate.
    .. versionadded:: 3.13
 
 
+.. c:function:: int PyLong_Sign(PyObject *obj)
+
+   Return ``-1``, ``0`` or ``+1`` if the integer object *obj* is negative, zero
+   or positive, respectively.
+
+   Returns ``-1`` on error.  Use :c:func:`PyErr_Occurred` to disambiguate.
```

Comment:
This API is bad. Please change the public API to: `int PyLong_Sign(PyObject *obj, int *result)`.

* Set `*result` and return 0 on success.
* Return -1 on error.

This way, we don't have to check for PyErr_Occurred().

See C API guidelines: https://devguide.python.org/developer-workflow/c-api/index.html#guidelines-for-expanding-changing-the-public-api

Reply:
Hmm, I'm not sure that this is against the C API guidline.  ("Most other API must return -1 with an exception raised on error, and 0 on success.", etc)

With the new design, 1) ``_PyLong_Sign()`` should be restored, 2) we loose ability to do something like:
```c
if (PyLong_Sign(its_surely_a_long_int) < 0) {
    <do stuff>;
}
```
Maybe we should instead document all cases, when an exception could be raised?  (It's easy: ``NULL`` pointer or not an ``int`` subclass.)

In this way, users could choose if they actually want to call ``PyErr_Occurred()`` or be sure this can't happen in their code.  Most consumers of the API (e.g. gmpy2), probably, will fall to the second camp (and just remove ``_`` prefix).

Reply:
> Most other API must return -1 with an exception raised on error, and 0 on success.

Yes, new API functions have to follow this guideline. The result is an argument passed by pointer:

```c
int sign;
if (PyLong_Sign(obj, &sign) < 0) { ... handle error ... }
... use sign ...
```

> Maybe we should instead document all cases, when an exception could be raised? (It's easy: NULL pointer or not an int subclass.)

This code fails obviously (go to "handle error") if obj is not an integer. For example, if obj is a string.

Usually, we don't document error cases, since it would hard to maintain this in the documentation. It's better to document the case which don't fail, like "obj must be an Python int object".

> In this way, users could choose if they actually want to call PyErr_Occurred() or be sure this can't happen in their code.

The chosen guideline is that all C API consumers must check for errors. If you know for sure that the function **cannot** fail, just ignore the function result and use `sign`.

Reply:
The problem of the current API is that check for `PyErr_Occurred()` is inefficient, whereas if we design the API differently, `PyErr_Occurred()` doesn't have to be called: it's faster.

Reply:
I hope, I did this.